scipy/special/_basic.py

```diff
@@ -2353,14 +2359,15 @@ def factorial(n, exact=False):
                     current = un[i + 1]
                     val *= _range_prod(prev, current)
                     out[n == current] = val
+
+            if np.isnan(n).any():
+                out = out.astype(np.float64)
+                out[np.isnan(n)] = n[np.isnan(n)]
             return out
     else:
-        if np.ndim(n) == 0:
-            return 0 if n < 0 else gamma(n + 1)
-
         n = asarray(n)
```

Comment:
Now you can delete the `assaray` line. Really this branch can just be `return ufuncs._factorial(n)`.